lib/expression/v2.ex

```diff
@@ -73,28 +76,80 @@ defmodule Expression.V2 do
     end
   end
 
-  @spec eval(String.t(), context :: Context.t()) :: [term]
-  def eval(expression, context \\ Context.new()) do
-    parts = compile(expression)
+  @doc """
+  Parse a string with an expression block into AST for the compile step
+  """
+  @spec parse_block(String.t()) ::
+          {:ok, [term]} | {:error, reason :: String.t(), bad_parts :: String.t()}
+  def parse_block(expression_block) do
+    case Parser.expression(expression_block) do
+      {:ok, ast, "", _, _, _} -> {:ok, ast}
+      {:error, _ast, remaining, _, _, _} -> {:error, "Unable to parse remainder", remaining}
+    end
+  end
+
+  @doc """
+  Evaluate a string with an expression block against a context
+  """
+  @spec eval_block(String.t(), context :: Context.t()) :: term
+  def eval_block(expression_block, context \\ Context.new()) do
+    {:ok, ast} = parse_block(expression_block)
+    function = Compile.compile(ast)
+    function.(context)
+  end
 
-    Enum.map(parts, fn
-      part when is_list(part) ->
-        part
-        |> Enum.map(&eval_part(&1, context))
-        |> hd()
+  @doc """
+  Evaluate a string with expressions against a given context
+  """
+  @spec eval(String.t(), context :: Context.t()) :: [term]
+  def eval(expression, context \\ Context.new()) when is_binary(expression) do
+    ast = compile(expression)
 
-      other ->
-        other
+    Enum.map(ast, fn
+      part when is_list(part) -> Enum.map(part, &eval_in_context(&1, context)) |> hd()
+      other -> other
```

Comment:
just a small optional change:

```suggestion
    expression
    |> compile()
    |> Enum.map(ast, fn
      part when is_list(part) -> Enum.map(part, &eval_in_context(&1, context)) |> List.first()
      other -> other
    end)
```

Reply:
done in 69d0c1e